### PR TITLE
Add Public Key Infra SONiC module

### PR DESCRIPTION
**Pull request by seiferteric**
##### SUMMARY
Add ansible support for SONiC PKI (Public Key Infra) model. This includes security-profiles and trust-stores for managing service certificates and related settings.

##### GitHub Issues
List the GitHub issues impacted by this PR. If no Github issues are affected, please indicate this with "N/A".

| GitHub Issue # |
| -------------- |
| N/A|

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME

sonic_pki

##### OUTPUT

```paste below

```
##### Regression Run Output

[regression-2023-09-15-15-59-38.html.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/12621665/regression-2023-09-15-15-59-38.html.pdf)

##### Detailed Regression Run Log

[regression_output.log](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/12621670/regression_output.log)

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.
- [x] I have provided a summary for this PR in valid "fragment" file format in the "changelogs/fragments" directory of this repository branch. Reference : [Ansible Change Log Document](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)

##### How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Manual unit testing
- [x] Automated regression testing

### CODE REVIEW

File: plugins/modules/sonic_pki.py
Comment (stalabi1) [inline plugins/modules/sonic_pki.py:96]: Please add examples for all states

File: plugins/modules/sonic_pki.py
Comment (stalabi1) [inline plugins/modules/sonic_pki.py:41]:
```suggestion
version_added: 1.0.0
```
```suggestion
version_added: 2.3.0
```

File: plugins/modules/sonic_pki.py
Comment (stalabi1) [inline plugins/modules/sonic_pki.py:94]:
```suggestion
            elements: str
  state:
    description:
      - The state of the configuration after module completion.
    type: str
    choices: ['merged', 'deleted', 'replaced', 'overridden']
    default: merged
```

File: plugins/module_utils/network/sonic/argspec/pki/pki.py
Comment (stalabi1) [inline plugins/module_utils/network/sonic/argspec/pki/pki.py:48]:
```suggestion
                        'profile-name': {'required': True, 'type': 'str'},
```

File: plugins/module_utils/network/sonic/argspec/pki/pki.py
Comment (stalabi1) [inline plugins/module_utils/network/sonic/argspec/pki/pki.py:59]:
```suggestion
                        'name': {'required': True, 'type': 'str'}
```

**Comment by stalabi1** [comment]
You need to add the UT for this model in tests/unit/modules/network/sonic

**Comment by seiferteric** [comment]
@stalabi1 @kerry-meyer I believe I have addressed all the PR comments, please review.

File: changelogs/fragments/278-pki.yaml
Comment (stalabi1) [inline changelogs/fragments/278-pki.yaml:1]: Please rename the yaml '280-pki'. The number in the name should be the PR #.

File: plugins/modules/sonic_pki.py
Comment (stalabi1) [reply plugins/modules/sonic_pki.py:102]: Please also add examples for overridden and replaced.

File: changelogs/fragments/278-pki.yaml
Comment (seiferteric) [reply changelogs/fragments/278-pki.yaml:1]: done

File: plugins/modules/sonic_pki.py
Comment (seiferteric) [reply plugins/modules/sonic_pki.py:102]: added

File: plugins/modules/sonic_pki.py
Comment (stalabi1) [inline plugins/modules/sonic_pki.py:59]:
```suggestion
            description: The name of the Trust Store
             required: True
```

File: plugins/modules/sonic_pki.py
Comment (stalabi1) [inline plugins/modules/sonic_pki.py:71]:
```suggestion
            description: Profile Name
            required: True
```

**Review by stalabi1 — APPROVED** [review]

File: tests/unit/modules/network/sonic/fixtures/sonic_pki.yaml
Comment (kerry-meyer) [inline tests/unit/modules/network/sonic/fixtures/sonic_pki.yaml:1]: The UT coverage for this resource module is pretty far below our minimum target of 90%. (70% is allowable, but the current coverage is 62%.)

Please modify the UT test cases to allow traversal of more of the code.

By modifying the deletion case to delete specific attributes instead of doing a "delete all", it looks like you would be able to at least get the coverage above 70%.

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
I am posting some follow-up code review comments.

(Sorry for the delay in posting these comments.)

In addition to addressing (fixing or responding to) the minor inline comments and the comment on the needed enhancement for the unit test "fixtures" file, the Ansible sanity errors need to be resolved before this PR can be merged.

You can view the sanity results at the bottom of the "Conversation" tab for this PR by clicking on "Details" for any of the failing sanity tests. I think they are all the same.

One example URL containing the error list is:

https://github.com/ansible-collections/dellemc.enterprise_sonic/actions/runs/5525054308/job/15500167786?pr=280

UPDATE: All of the issued described above in this comment have been marked "Resolved".

**Review by stalabi1 — CHANGES_REQUESTED** [review]
Please delete the fragment file. It is no longer needed now.

File: plugins/module_utils/network/sonic/config/pki/pki.py
Comment (seiferteric) [reply plugins/module_utils/network/sonic/config/pki/pki.py:205]: have_dict is just a convenience, have has a list of security-profiles/trust-stores, so I would have to search through the list and find the matching one. Instead I just create this have_dict that has a dictionary of security-profiles/trust-stores that has the key as the name so I don't have to search for it.

**Comment by kerry-meyer** [comment]
Please fix the sanity errors flagged for this change set at the bottom of the PR "home" page. (There is a previous, unresolved comment about this explaining how to check them.) Based on a quick look at the failures, at least the following items need to be fixed to obtain a passing sanity result:
 - The argument names need to use "_" instead of "-" as a separator.
 - The argspec the "module" file definitions should match for all arguments. There are sanity failures for arguments marked as "required" in the argspec, but not in the module definition file (plugins/modules/sonic_pki.py).
 - The required format for the "author" field in the Documentation block of the "module" file is:
   author: "Eric Seifert (@seiferteric)".
- There is a stray "---" line at line 111 of the module file. Please delete it.
- In the "RETURN" block starting at line 267 of the module file, the following line is missing from both the "before" and "after" sections (immediately after each of the "returned: ..." lines, before each of the "sample: ..." lines):
      type: dict

File: tests/unit/modules/network/sonic/fixtures/sonic_pki.yaml
Comment (seiferteric) [reply tests/unit/modules/network/sonic/fixtures/sonic_pki.yaml:1]: Where are you seeing this? I ran `ansible-test units --coverage --requirements` followed by `ansible-test coverage report` and for the pki line I see:

```
tests/unit/modules/network/sonic/test_sonic_pki.py                                         47      0      2      0   100%
```

Comment (kerry-meyer) [reply tests/unit/modules/network/sonic/fixtures/sonic_pki.yaml:1]: The output shown on your preceding comment indicates that 100% of the UT tests passed. This means that the real output completely matched the expected output for all cases, and confirms that the logic tested by the currently existing UT cases is functioning exactly as required. (So far, so good.)

The coverage number mentioned in my comments is a measurement of the proportion of the existing instructions that are actually being traversed by the currently defined UT test cases.. This is the place where the coverage percentage is lagging below the 70% minimum. (The primary code file of concern is the "config" file for the resource module (plugins/module_utils/network/sonic/config/pki/pki.py).

To view the code coverage results:

- Click on the "Details" link to the right of the "Code coverage ..." line in the test result summary section at the bottom of the PR "home" page.
- Expand the "Run code coverage tests" section of the results that appear by clicking on the ">" next to that section title.
- Scroll down in the displayed list of files to the line containing the results for the config file for this resource module ()plugins/module_utils/network/sonic/config/pki/pki.py). The "coverage" column (far right column) shows the percentage for code coverage of the instructions in this file during running of the UT "coverage" tests.

**Review by stalabi1 — APPROVED** [review]
Thanks for addressing my review comments.

File: tests/regression/roles/sonic_pki/defaults/main.yml
Comment (kerry-meyer) [inline tests/regression/roles/sonic_pki/defaults/main.yml:26]: Please expand this test case or add another one to test the following list expansion cases:

- Adding one or more ca_names to an existing trust_store ca_name list.
- Adding one or more URLs to an existing security profile ocsp_responder list.
- Adding another trust store and/or security profile.

File: tests/regression/roles/sonic_pki/defaults/main.yml
Comment (kerry-meyer) [inline tests/regression/roles/sonic_pki/defaults/main.yml:35]: In addition to testing deletion of a list item, please expand this test case to cover deletion of other (non-list_ attributes (e.g. a security profile "trust_store").

File: tests/regression/roles/sonic_pki/defaults/main.yml
Comment (kerry-meyer) [inline tests/regression/roles/sonic_pki/defaults/main.yml:65]: This is a good test case, but the code needs to be changed to handle it in a way that replaces the entire security profile with the new content. (A corresponding comment has been posted inline in the "config" file diffs.)

File: tests/regression/roles/sonic_pki/defaults/main.yml
Comment (kerry-meyer) [inline tests/regression/roles/sonic_pki/defaults/main.yml:82]:
```suggestion
    description: Override parameter of security_profiles and trust_stores
```

File: tests/regression/roles/sonic_pki/defaults/main.yml
Comment (kerry-meyer) [inline tests/regression/roles/sonic_pki/defaults/main.yml:81]: Please modify this test case and the preceding "merged" test to enable verification that specifying "overridden" state for one or more trust stores and or security profiles results in deletion of all existing trust stores and security profiles that are not specified in the "overridden" attributes.

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
I have submitted additional comments in several of the files. None of the comments request any significant structural changes, but there are a number of additional/expanded integration test cases requested.

File: tests/unit/modules/network/sonic/fixtures/sonic_pki.yaml
Comment (kerry-meyer) [reply tests/unit/modules/network/sonic/fixtures/sonic_pki.yaml:1]: Coverage has been increased to 70%.

Resolved.

File: plugins/module_utils/network/sonic/config/pki/pki.py
Comment (seiferteric) [reply plugins/module_utils/network/sonic/config/pki/pki.py:387]: Is there something you saw that didn't work? My testing shows it working correctly, so not sure why this change is needed.

Comment (seiferteric) [reply plugins/module_utils/network/sonic/config/pki/pki.py:391]: okay added.

**Review by kerry-meyer — COMMENTED** [review]
I have checked all of the incremental changes you have posted to this PR.

All of them look fine, and I have marked all of the issues "Resolved". 

Thanks for making these changes.

Please re-run the regression test suite with the current set of tests and code changes and post the results.

I will approve the PR and merge it after verifying that the test cases run successfully (and yield the correct results).

**Review by kerry-meyer — APPROVED** [review]
Thanks for posting the latest test results.

All of the test results look good.

Approved.

**Pull request merged**
